File: bindings/python/test_parity.py
import json
import os
import unittest
from pathlib import Path

import stdbr

GOLDEN_PATH = os.environ.get("GOLDEN_JSON") or str(
    Path(__file__).resolve().parent.parent.parent / "tests" / "parity" / "golden.json"
)
GOLDEN = json.loads(Path(GOLDEN_PATH).read_text())


class TestCpf(unittest.TestCase):
    def test_parse(self):
        for c in GOLDEN["cpf"]["parse"]:
            with self.subTest(input=c["input"]):
                cpf = stdbr.Cpf.parse(c["input"])
                self.assertEqual(cpf.as_str(), c["digits_only"])
                self.assertEqual(cpf.formatted(), c["formatted"])
                self.assertEqual(cpf.masked(), c["masked"])
                self.assertEqual(int(cpf.fiscal_region), c["fiscal_region"])
                self.assertEqual(cpf.check_digits[0], c["check_digits"][0])
                self.assertEqual(cpf.check_digits[1], c["check_digits"][1])

    def test_is_valid(self):
        for c in GOLDEN["cpf"]["is_valid"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cpf_is_valid(c["input"]), c["expected"])

    def test_is_valid_strict(self):
        for c in GOLDEN["cpf"]["is_valid_strict"]:
            with self.subTest(input=c["input"]):
                if c["valid"]:
                    stdbr.cpf_is_valid_strict(c["input"])
                else:
                    with self.assertRaises(ValueError) as ctx:
                        stdbr.cpf_is_valid_strict(c["input"])
                    self.assertEqual(str(ctx.exception), c["error"])

    def test_format(self):
        for c in GOLDEN["cpf"]["format"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cpf_format(c["input"]), c["expected"])

    def test_remove_symbols(self):
        for c in GOLDEN["cpf"]["remove_symbols"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cpf_remove_symbols(c["input"]), c["expected"])

    def test_compute_check_digits(self):
        for c in GOLDEN["cpf"]["compute_check_digits"]:
            with self.subTest(base=c["base"]):
                result = stdbr.cpf_compute_check_digits(c["base"])
                if c["expected"] is None:
                    self.assertIsNone(result)
                else:
                    self.assertEqual(result[0], c["expected"][0])
                    self.assertEqual(result[1], c["expected"][1])

    def test_generate_roundtrip(self):
        raw = stdbr.cpf_generate()
        self.assertTrue(stdbr.cpf_is_valid(raw))
        self.assertEqual(stdbr.Cpf.parse(raw).as_str(), raw)


class TestCnpj(unittest.TestCase):
    def test_parse(self):
        for c in GOLDEN["cnpj"]["parse"]:
            with self.subTest(input=c["input"]):
                cnpj = stdbr.Cnpj.parse(c["input"])
                self.assertEqual(cnpj.as_str(), c["digits_only"])
                self.assertEqual(cnpj.formatted(), c["formatted"])
                self.assertEqual(cnpj.masked(), c["masked"])
                self.assertEqual(cnpj.raiz(), c["raiz"])
                self.assertEqual(cnpj.ordem(), c["ordem"])
                self.assertEqual(int(cnpj.kind), c["kind"])
                self.assertEqual(int(cnpj.establishment_type), c["establishment_type"])
                self.assertEqual(cnpj.check_digits[0], c["check_digits"][0])
                self.assertEqual(cnpj.check_digits[1], c["check_digits"][1])

    def test_is_valid(self):
        for c in GOLDEN["cnpj"]["is_valid"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cnpj_is_valid(c["input"]), c["expected"])

    def test_is_valid_strict(self):
        for c in GOLDEN["cnpj"]["is_valid_strict"]:
            with self.subTest(input=c["input"]):
                if c["valid"]:
                    stdbr.cnpj_is_valid_strict(c["input"])
                else:
                    with self.assertRaises(ValueError) as ctx:
                        stdbr.cnpj_is_valid_strict(c["input"])
                    self.assertEqual(str(ctx.exception), c["error"])

    def test_format(self):
        for c in GOLDEN["cnpj"]["format"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cnpj_format(c["input"]), c["expected"])

    def test_remove_symbols(self):
        for c in GOLDEN["cnpj"]["remove_symbols"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cnpj_remove_symbols(c["input"]), c["expected"])

    def test_compute_check_digits(self):
        for c in GOLDEN["cnpj"]["compute_check_digits"]:
            with self.subTest(base=c["base"]):
                result = stdbr.cnpj_compute_check_digits(c["base"])
                if c["expected"] is None:
                    self.assertIsNone(result)
                else:
                    self.assertEqual(result[0], c["expected"][0])
                    self.assertEqual(result[1], c["expected"][1])

    def test_generate_roundtrip(self):
        raw = stdbr.cnpj_generate(stdbr.CnpjKind.Numeric)
        self.assertTrue(stdbr.cnpj_is_valid(raw))
        self.assertEqual(stdbr.Cnpj.parse(raw).as_str(), raw)


class TestCep(unittest.TestCase):
    def test_parse(self):
        for c in GOLDEN["cep"]["parse"]:
            with self.subTest(input=c["input"]):
                cep = stdbr.Cep.parse(c["input"])
                self.assertEqual(cep.as_str(), c["digits_only"])
                self.assertEqual(cep.formatted(), c["formatted"])
                self.assertEqual(cep.masked(), c["masked"])
                self.assertEqual(int(cep.postal_region), c["postal_region"])
                state = cep.state
                abbr = stdbr.state_abbreviation(state) if state is not None else None
                self.assertEqual(abbr, c["state"])

    def test_is_valid(self):
        for c in GOLDEN["cep"]["is_valid"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cep_is_valid(c["input"]), c["expected"])

    def test_is_valid_strict(self):
        for c in GOLDEN["cep"]["is_valid_strict"]:
            with self.subTest(input=c["input"]):
                if c["valid"]:
                    stdbr.cep_is_valid_strict(c["input"])
                else:
                    with self.assertRaises(ValueError) as ctx:
                        stdbr.cep_is_valid_strict(c["input"])
                    self.assertEqual(str(ctx.exception), c["error"])

    def test_format(self):
        for c in GOLDEN["cep"]["format"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cep_format(c["input"]), c["expected"])

    def test_remove_symbols(self):
        for c in GOLDEN["cep"]["remove_symbols"]:
            with self.subTest(input=c["input"]):
                self.assertEqual(stdbr.cep_remove_symbols(c["input"]), c["expected"])

    def test_generate_roundtrip(self):
        raw = stdbr.cep_generate()
        self.assertTrue(stdbr.cep_is_valid(raw))
        self.assertEqual(stdbr.Cep.parse(raw).as_str(), raw)


class TestUf(unittest.TestCase):
    def test_states(self):
        states = stdbr.all_states()
        self.assertEqual(len(states), len(GOLDEN["uf"]["states"]))
        for i, g in enumerate(GOLDEN["uf"]["states"]):
            with self.subTest(i=i):
                self.assertEqual(stdbr.state_abbreviation(states[i]), g["abbreviation"])
                self.assertEqual(stdbr.state_name(states[i]), g["name"])

    def test_from_abbreviation(self):
        for c in GOLDEN["uf"]["from_abbreviation"]:
            with self.subTest(input=c["input"]):
                s = stdbr.state_from_abbreviation(c["input"])
                result = stdbr.state_abbreviation(s) if s is not None else None
                self.assertEqual(result, c["expected"])


class TestMunicipio(unittest.TestCase):
    def test_count(self):
        self.assertEqual(stdbr.municipio_count(), GOLDEN["municipio"]["count"])

    def test_from_ibge_code(self):
        for c in GOLDEN["municipio"]["from_ibge_code"]:
            with self.subTest(code=c["code"]):
                m = stdbr.municipio_from_ibge_code(c["code"])
                if c.get("expected") is None and "name" not in c:
                    self.assertIsNone(m)
                else:
                    self.assertEqual(m.name, c["name"])
                    self.assertEqual(stdbr.state_abbreviation(m.state), c["state"])
                    self.assertEqual(m.is_capital, c["is_capital"])

    def test_capital_of(self):
        for c in GOLDEN["municipio"]["capital_of"]:
            with self.subTest(state=c["state"]):
                cap = stdbr.municipio_capital_of(stdbr.state_from_abbreviation(c["state"]))
                self.assertEqual(cap.name, c["name"])
                self.assertEqual(cap.ibge_code, c["ibge_code"])

    def test_search_by_name(self):
        for s in GOLDEN["municipio"]["search_by_name"]:
            with self.subTest(query=s["query"]):
                results = stdbr.municipio_search_by_name(s["query"])
                self.assertEqual(len(results), len(s["results"]))
                for i, e in enumerate(s["results"]):
                    self.assertEqual(results[i].ibge_code, e["ibge_code"])
                    self.assertEqual(results[i].name, e["name"])

    def test_by_state_count(self):
        for c in GOLDEN["municipio"]["by_state_count"]:
            with self.subTest(state=c["state"]):
                st = stdbr.state_from_abbreviation(c["state"])
                self.assertEqual(len(stdbr.municipios_by_state(st)), c["count"])


if __name__ == "__main__":
    unittest.main()

File: bindings/nodejs/test_parity.js
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { createRequire } from "module";
import { dirname, join } from "path";
import { describe, it } from "node:test";
import { strictEqual, ok } from "node:assert";

const __dirname = dirname(fileURLToPath(import.meta.url));
const goldenPath = process.env.GOLDEN_JSON || join(__dirname, "../../tests/parity/golden.json");
const golden = JSON.parse(readFileSync(goldenPath, "utf8"));

const require = createRequire(import.meta.url);
const stdbr = require("./stdbr.node");
const {
  Cpf, cpfIsValid, cpfIsValidStrict, cpfFormat, cpfRemoveSymbols,
  cpfGenerate, cpfComputeCheckDigits,
  Cnpj, cnpjIsValid, cnpjIsValidStrict, cnpjFormat, cnpjRemoveSymbols,
  cnpjGenerate, cnpjComputeCheckDigits, CnpjKind,
  Cep, cepIsValid, cepIsValidStrict, cepFormat, cepRemoveSymbols,
  cepGenerate,
  stateAbbreviation, stateName, stateFromAbbreviation, allStates,
  municipioFromIbgeCode, municipioCapitalOf, municipiosByState,
  municipioSearchByName, municipioCount,
} = stdbr;

function strictTest(validateFn, cases) {
  for (const c of cases) {
    it(`${validateFn.name}(${c.input})`, () => {
      if (c.valid) {
        validateFn(c.input);
      } else {
        try {
          validateFn(c.input);
          ok(false, "should throw");
        } catch (e) {
          strictEqual(e.message, c.error);
        }
      }
    });
  }
}

function checkDigitsTest(computeFn, cases) {
  for (const c of cases) {
    it(`${computeFn.name}(${c.base})`, () => {
      const result = computeFn(c.base);
      if (c.expected === null) {
        strictEqual(result, null);
      } else {
        strictEqual(result[0], c.expected[0]);
        strictEqual(result[1], c.expected[1]);
      }
    });
  }
}

describe("CPF", () => {
  describe("parse", () => {
    for (const c of golden.cpf.parse) {
      it(c.input, () => {
        const cpf = Cpf.parse(c.input);
        strictEqual(cpf.asStr(), c.digits_only);
        strictEqual(cpf.formatted(), c.formatted);
        strictEqual(cpf.masked(), c.masked);
        strictEqual(cpf.fiscalRegion, c.fiscal_region);
        strictEqual(cpf.checkDigits[0], c.check_digits[0]);
        strictEqual(cpf.checkDigits[1], c.check_digits[1]);
      });
    }
  });

  describe("is_valid", () => {
    for (const c of golden.cpf.is_valid)
      it(c.input || "(empty)", () => strictEqual(cpfIsValid(c.input), c.expected));
  });

  describe("is_valid_strict", () => strictTest(cpfIsValidStrict, golden.cpf.is_valid_strict));

  describe("format", () => {
    for (const c of golden.cpf.format)
      it(c.input || "(empty)", () => strictEqual(cpfFormat(c.input) ?? null, c.expected));
  });

  describe("remove_symbols", () => {
    for (const c of golden.cpf.remove_symbols)
      it(c.input, () => strictEqual(cpfRemoveSymbols(c.input), c.expected));
  });

  describe("compute_check_digits", () => checkDigitsTest(cpfComputeCheckDigits, golden.cpf.compute_check_digits));

  it("generate roundtrip", () => {
    const raw = cpfGenerate();
    ok(cpfIsValid(raw));
    strictEqual(Cpf.parse(raw).asStr(), raw);
  });
});

describe("CNPJ", () => {
  describe("parse", () => {
    for (const c of golden.cnpj.parse) {
      it(c.input, () => {
        const cnpj = Cnpj.parse(c.input);
        strictEqual(cnpj.asStr(), c.digits_only);
        strictEqual(cnpj.formatted(), c.formatted);
        strictEqual(cnpj.masked(), c.masked);
        strictEqual(cnpj.raiz(), c.raiz);
        strictEqual(cnpj.ordem(), c.ordem);
        strictEqual(cnpj.kind, c.kind);
        strictEqual(cnpj.establishmentType, c.establishment_type);
        strictEqual(cnpj.checkDigits[0], c.check_digits[0]);
        strictEqual(cnpj.checkDigits[1], c.check_digits[1]);
      });
    }
  });

  describe("is_valid", () => {
    for (const c of golden.cnpj.is_valid)
      it(c.input || "(empty)", () => strictEqual(cnpjIsValid(c.input), c.expected));
  });

  describe("is_valid_strict", () => strictTest(cnpjIsValidStrict, golden.cnpj.is_valid_strict));

  describe("format", () => {
    for (const c of golden.cnpj.format)
      it(c.input || "(empty)", () => strictEqual(cnpjFormat(c.input) ?? null, c.expected));
  });

  describe("remove_symbols", () => {
    for (const c of golden.cnpj.remove_symbols)
      it(c.input, () => strictEqual(cnpjRemoveSymbols(c.input), c.expected));
  });

  describe("compute_check_digits", () => checkDigitsTest(cnpjComputeCheckDigits, golden.cnpj.compute_check_digits));

  it("generate roundtrip", () => {
    const raw = cnpjGenerate(CnpjKind.Numeric);
    ok(cnpjIsValid(raw));
    strictEqual(Cnpj.parse(raw).asStr(), raw);
  });
});

describe("CEP", () => {
  describe("parse", () => {
    for (const c of golden.cep.parse) {
      it(c.input, () => {
        const cep = Cep.parse(c.input);
        strictEqual(cep.asStr(), c.digits_only);
        strictEqual(cep.formatted(), c.formatted);
        strictEqual(cep.masked(), c.masked);
        strictEqual(cep.postalRegion, c.postal_region);
        const abbr = cep.state != null ? stateAbbreviation(cep.state) : null;
        strictEqual(abbr, c.state);
      });
    }
  });

  describe("is_valid", () => {
    for (const c of golden.cep.is_valid)
      it(c.input || "(empty)", () => strictEqual(cepIsValid(c.input), c.expected));
  });

  describe("is_valid_strict", () => strictTest(cepIsValidStrict, golden.cep.is_valid_strict));

  describe("format", () => {
    for (const c of golden.cep.format)
      it(c.input || "(empty)", () => strictEqual(cepFormat(c.input) ?? null, c.expected));
  });

  describe("remove_symbols", () => {
    for (const c of golden.cep.remove_symbols)
      it(c.input, () => strictEqual(cepRemoveSymbols(c.input), c.expected));
  });

  it("generate roundtrip", () => {
    const raw = cepGenerate();
    ok(cepIsValid(raw));
    strictEqual(Cep.parse(raw).asStr(), raw);
  });
});

describe("UF", () => {
  it("states", () => {
    const states = allStates();
    strictEqual(states.length, golden.uf.states.length);
    for (let i = 0; i < golden.uf.states.length; i++) {
      strictEqual(stateAbbreviation(states[i]), golden.uf.states[i].abbreviation);
      strictEqual(stateName(states[i]), golden.uf.states[i].name);
    }
  });

  describe("from_abbreviation", () => {
    for (const c of golden.uf.from_abbreviation) {
      it(c.input || "(empty)", () => {
        const s = stateFromAbbreviation(c.input);
        strictEqual(s != null ? stateAbbreviation(s) : null, c.expected);
      });
    }
  });
});

describe("Municipio", () => {
  it("count", () => strictEqual(municipioCount(), golden.municipio.count));

  describe("from_ibge_code", () => {
    for (const c of golden.municipio.from_ibge_code) {
      it(String(c.code), () => {
        const m = municipioFromIbgeCode(c.code);
        if (c.expected === null) {
          strictEqual(m, null);
        } else {
          strictEqual(m.name, c.name);
          strictEqual(stateAbbreviation(m.state), c.state);
          strictEqual(m.isCapital, c.is_capital);
        }
      });
    }
  });

  describe("capital_of", () => {
    for (const c of golden.municipio.capital_of) {
      it(c.state, () => {
        const cap = municipioCapitalOf(stateFromAbbreviation(c.state));
        strictEqual(cap.name, c.name);
        strictEqual(cap.ibgeCode, c.ibge_code);
      });
    }
  });

  describe("search_by_name", () => {
    for (const s of golden.municipio.search_by_name) {
      it(s.query, () => {
        const results = municipioSearchByName(s.query);
        strictEqual(results.length, s.results.length);
        for (let i = 0; i < s.results.length; i++) {
          strictEqual(results[i].ibgeCode, s.results[i].ibge_code);
          strictEqual(results[i].name, s.results[i].name);
        }
      });
    }
  });

  describe("by_state_count", () => {
    for (const c of golden.municipio.by_state_count) {
      it(c.state, () => {
        strictEqual(municipiosByState(stateFromAbbreviation(c.state)).length, c.count);
      });
    }
  });
});
